File: src/main/java/net/zenxarch/bot/process/CraftProcess.java
package net.zenxarch.bot.process;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.function.Predicate;
import net.minecraft.client.MinecraftClient;
import net.minecraft.client.gui.screen.ingame.CraftingScreen;
import net.minecraft.client.gui.screen.ingame.FurnaceScreen;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.item.Items;
import net.minecraft.recipe.Recipe;
// import net.minecraft.recipe.BlastingRecipe;
import net.minecraft.recipe.ShapedRecipe;
import net.minecraft.recipe.SmeltingRecipe;
// import net.minecraft.recipe.SmokingRecipe;
import net.minecraft.screen.FurnaceScreenHandler;
import net.minecraft.screen.slot.SlotActionType;
import net.minecraft.util.Pair;
import net.zenxarch.bot.util.RecipeUtil;

public final class CraftProcess {
  private static ArrayList<Pair<ShapedRecipe, Integer>>
      craftingQueue = new ArrayList<>();
  private static ArrayList<Pair<SmeltingRecipe, Integer>>
      smeltingQueue = new ArrayList<>();
  // private static ArrayList<BlastingRecipe> blastingQueue =
  //   new ArrayList<>();
  // private static ArrayList<SmokingRecipe> smokingQueue =
  //   new ArrayList<>();
  private static final MinecraftClient mc =
      MinecraftClient.getInstance();
  public static void preTick() {
    if (craftingQueue.size() > 0 && mc.currentScreen instanceof
                                        CraftingScreen) {
      craft();
    }
    if (smeltingQueue.size() > 0 && mc.currentScreen instanceof
                                        FurnaceScreen) {
      smelt();
    }
  }

  public static int checkMaxCraftable(Recipe<?> r) {
    HashMap<List<Item>, Integer> sr = RecipeUtil.simpleIngredient(r);
    var inv = mc.player.getInventory()
                  .main.stream()
                  .filter(Predicate.not(ItemStack::isEmpty))
                  .toList();
    var itemAvail = new HashMap<List<Item>, Integer>();
    for (var is : inv) {
      for (var k : sr.keySet()) {
        if (k.contains(is.getItem()))
          itemAvail.put(k,
                        itemAvail.getOrDefault(k, 0) + is.getCount());
      }
    }
    for (var e : sr.entrySet()) {
      itemAvail.put(e.getKey(),
                    itemAvail.get(e.getKey()) / e.getValue());
    }
    return itemAvail.values().stream().min(Integer::compare).get();
  }

  private static void craft() {
    var r = craftingQueue.get(0);
    var cs = (CraftingScreen)mc.currentScreen;
    var csh = cs.getScreenHandler();
    var im = mc.interactionManager;
    var os = csh.getCraftingResultSlotIndex();
    if (csh.getSlot(os).getStack().equals(r.getLeft().getOutput())) {
      im.clickSlot(csh.syncId, os, 0, SlotActionType.QUICK_MOVE,
                   mc.player);
      r.setRight(r.getRight() - 1);
      if (r.getRight() == 0) {
        craftingQueue.remove(0);
      }
      return;
    }
    im.clickRecipe(csh.syncId, r.getLeft(), false);
  }

  private static void smelt() {
    var r = smeltingQueue.get(0).getLeft();
    var cs = (FurnaceScreen)mc.currentScreen;
    var csh = cs.getScreenHandler();
    var os = 2;
    // var is = 0;
    // var fs = 1;
    var im = mc.interactionManager;
    if (csh.getSlot(0).getStack().equals(r.getOutput())) {
      im.clickSlot(csh.syncId, os, 0, SlotActionType.QUICK_MOVE,
                   mc.player);
      var q = smeltingQueue.get(0);
      q.setRight(q.getRight() - 1);
      if (q.getRight() == 0) {
        smeltingQueue.remove(0);
      }
      return;
    }
    if (!checkSmeltInput(csh, r)) {
      im.clickRecipe(csh.syncId, r, false);
    }
    if (csh.isBurning()) {
      return;
    }
    for (int i = csh.slots.size() - 9 - 27; i < csh.slots.size();
         i++) {
      if (!csh.getSlot(i).getStack().isOf(Items.COAL)) {
        continue;
      }
      im.clickSlot(csh.syncId, i, 0, SlotActionType.QUICK_MOVE,
                   mc.player);
      break;
    }
  }

  private static boolean checkSmeltInput(FurnaceScreenHandler csh,
                                         SmeltingRecipe r) {
    assert r.getIngredients().size() == 1;
    return r.getIngredients().get(0).test(csh.getSlot(0).getStack());
  }

  public static void postTick() {}

  public static void enqueue(ShapedRecipe r, Integer amt) {
    craftingQueue.add(new Pair<ShapedRecipe, Integer>(r, amt));
  }

  public static void enqueue(SmeltingRecipe r, Integer amt) {
    smeltingQueue.add(new Pair<SmeltingRecipe, Integer>(r, amt));
  }

  /*

  public static void enqueue(BlastingRecipe r){
    blastingQueue.add(r);
  }

  public static void enqueue(SmokingRecipe r){
    smokingQueue.add(r);
  }

  */
}

File: src/main/scala/net/zenxarch/bot/process/CraftProcess.scala
package net.zenxarch.bot.process

import net.minecraft.client.MinecraftClient
import net.minecraft.client.gui.screen.ingame.CraftingScreen
import net.minecraft.client.gui.screen.ingame.FurnaceScreen
import net.minecraft.item.Item
import net.minecraft.item.ItemStack
import net.minecraft.item.Items
import net.minecraft.recipe.Recipe
// import net.minecraft.recipe.BlastingRecipe
import net.minecraft.recipe.ShapedRecipe
import net.minecraft.recipe.SmeltingRecipe
// import net.minecraft.recipe.SmokingRecipe
import net.minecraft.screen.FurnaceScreenHandler
import net.minecraft.screen.slot.SlotActionType
import net.minecraft.util.Pair
import net.zenxarch.bot.util.RecipeUtil
import net.zenxarch.bot.ZenBot.mc
import scala.collection.mutable.HashMap
import scala.collection.mutable.ListBuffer
import net.minecraft.client.network.ClientPlayerEntity
import scala.jdk.CollectionConverters._
import baritone.aa

object CraftProcess {
  private var craftingQueue =
    new ListBuffer[(ShapedRecipe, Int)]()
  private var smeltingQueue =
    new ListBuffer[(SmeltingRecipe, Int)]()
  // private static ArrayList<BlastingRecipe> blastingQueue =
  //   new ArrayList<>()
  // private static ArrayList<SmokingRecipe> smokingQueue =
  //   new ArrayList<>()
  def preTick() = {
    if (
      craftingQueue.length > 0 && mc.currentScreen.isInstanceOf[CraftingScreen]
    ) {
      craft()
    }
    if (
      smeltingQueue.length > 0 && mc.currentScreen.isInstanceOf[FurnaceScreen]
    ) {
      smelt()
    }
  }

  def checkMaxCraftable(r: Recipe[?]): Int = {
    val sr = RecipeUtil.simpleIngredient(r)
    val p: ClientPlayerEntity = mc.player
    var inv = p
      .getInventory()
      .main
      .stream()
      .filter(s => !s.isEmpty())
      .toList()
      .asScala
    var itemAvail = new HashMap[List[Item], Integer]()
    for (is <- inv) {
      for (k <- sr.keys) {
        if (k.contains(is.getItem()))
          itemAvail.put(
            k,
            itemAvail.getOrElse(k, 0).asInstanceOf[Int] + is.getCount()
          )
      }
    }
    for ((k, v) <- sr) {
      itemAvail.put(k, itemAvail.getOrElse(k, 0).asInstanceOf[Int] / v)
    }
    return itemAvail.reduceLeft((a, b) => if (a(1) < b(1)) a else b)(1)
  }

  private def craft(): Unit = {
    var r = craftingQueue(0)
    if (checkMaxCraftable(r(0)) == 0) {
      return
    }
    var cs = mc.currentScreen.asInstanceOf[CraftingScreen]
    var csh = cs.getScreenHandler()
    var im = mc.interactionManager
    var os = csh.getCraftingResultSlotIndex()
    if (
      csh.getSlot(os).getStack().getItem() ==
        r(0).getOutput().getItem()
    ) {
      im.clickSlot(csh.syncId, os, 0, SlotActionType.QUICK_MOVE, mc.player)
      craftingQueue(0) = r.copy(_2 = r(1) - 1)
      if (r(1) < 2) {
        craftingQueue.remove(0)
      }
      return
    }
    im.clickRecipe(csh.syncId, r(0), false)
  }

  private def smelt(): Unit = {
    var r = smeltingQueue(0)(0)
    var cs = mc.currentScreen.asInstanceOf[FurnaceScreen]
    var csh = cs.getScreenHandler()
    var os = 2
    // var is = 0
    // var fs = 1
    var im = mc.interactionManager
    if (csh.getSlot(0).getStack().equals(r.getOutput())) {
      im.clickSlot(csh.syncId, os, 0, SlotActionType.QUICK_MOVE, mc.player)
      val q = smeltingQueue(0)
      smeltingQueue(0) = q.copy(_2 = q(1) - 1)
      if (q(1) < 2) {
        smeltingQueue.remove(0)
      }
      return
    }
    if (!checkSmeltInput(csh, r)) {
      im.clickRecipe(csh.syncId, r, false)
    }
    if (csh.isBurning()) {
      return
    }
    for
      i <- csh.slots.size() - 9 - 27 until csh.slots.size()
      if csh.getSlot(i).getStack().isOf(Items.COAL)
    do
      im.clickSlot(csh.syncId, i, 0, SlotActionType.QUICK_MOVE, mc.player)
      return
  }

  private def checkSmeltInput(
      csh: FurnaceScreenHandler,
      r: SmeltingRecipe
  ): Boolean = {
    assert(r.getIngredients().size() == 1)
    return r.getIngredients().get(0).test(csh.getSlot(0).getStack())
  }

  def postTick() = {}

  def enqueue(r: ShapedRecipe, amt: Int) = {
    craftingQueue += ((r, amt))
  }

  def enqueue(r: SmeltingRecipe, amt: Int) = {
    smeltingQueue += ((r, amt))
  }

  /*

   static void enqueue(BlastingRecipe r){
    blastingQueue.add(r)
  }

   static void enqueue(SmokingRecipe r){
    smokingQueue.add(r)
  }

   */
}
